feat: "bring your own" embeddings models

jcorrie opened the pull request:
On top of the existing facility to load embedding models from the huggingface hub, this exposes an interface for constituting a UserDefinedEmbeddingModel by passing in the bytes of the relevant files. (This is preferred over accepting pathbuf arguments, as it gives the user flexibility to load files from any source.)

Anush008 commented:
Hi @jcorrie. Could you document this in the README.md for more visibility

Anush008 commented:
Hi @jcorrie. Could you please document this in the README.md for more visibility

jcorrie commented:
> Hi @jcorrie. Could you please document this in the README.md for more visibility

Hi @Anush008 - yes, will do. I'm also looking at the suggestions made by Sourcery.

jcorrie commented on src/lib.rs:
```diff
@@ -125,6 +127,24 @@
+pub struct InitOptionsUserDefined {
+    pub execution_providers: Vec<ExecutionProviderDispatch>,
+    pub max_length: usize,
+}
```
More descriptive messages to be added within TextEmbedding::load_tokenizer() .

jcorrie commented on src/lib.rs:
```diff
@@ -125,6 +127,24 @@ impl Default for InitOptions {
     }
 }
 
+/// Options for initializing UserDefinedEmbeddingModel
+///
+/// Model files are held by the UserDefinedEmbeddingModel struct
+#[derive(Debug, Clone)]
+pub struct InitOptionsUserDefined {
+    pub execution_providers: Vec<ExecutionProviderDispatch>,
+    pub max_length: usize,
+}
```
To be implemented as suggested.

jcorrie commented on src/lib.rs:
```diff
@@ -227,27 +292,40 @@
         embedding_model: &EmbeddingModel,
         cache_dir: &PathBuf,
-    ) -> Option<PathBuf> {
+    ) -> Result<PathBuf> {
```
Considered that existing error handling/error messages are sufficient.

jcorrie commented on src/lib.rs:
```diff
@@ -468,13 +546,20 @@
 
 /// Get the cached onnx file from the model directory
-fn get_cached_onnx_file(model: ModelInfo, cache_dir: &PathBuf) -> Option<PathBuf> {
+fn get_cached_onnx_file(model: ModelInfo, cache_dir: &PathBuf) -> Result<PathBuf> {
```
Considered that existing error handling/error messages are sufficient.

jcorrie commented on src/lib.rs:
```diff
@@ -468,13 +546,20 @@
 
 /// Get the cached onnx file from the model directory
-fn get_cached_onnx_file(model: ModelInfo, cache_dir: &PathBuf) -> Option<PathBuf> {
+fn get_cached_onnx_file(model: ModelInfo, cache_dir: &PathBuf) -> Result<PathBuf> {
```
Considered that existing error handling/error messages are sufficient.

jcorrie commented:
@Anush008  - I've added some blurb to the readme, but I didn't want to put too much detail in as to how a UserDefinedEmbeddingModel is constituted (it might be a bit noisy). I can put some more time into a demo if need be (I have some code from my working main.rs file I can repurpose).

Test also added for UserDefinedEmbeddingModel per sourcery suggestion - passes on my machine...  I didn't take on board all of their suggestions but I've left some comments if you wanted to take a look. Again, happy to put some more time into resolving those if you think helpful (I just thought some looked a bit like overkill).

Anush008 reviewed (approved):
Thank you.

Anush008 commented on README.md:
```diff
@@ -53,6 +55,11 @@ use fastembed::{TextEmbedding, InitOptions, EmbeddingModel};
 // With default InitOptions
 let model = TextEmbedding::try_new(Default::default())?;
 
+// Alternatively, here a "bring your own" model could be loaded
+// Users will need to manually pass in the bytes of the relevant model files
+// This includes  the .onnx file itself and json files to constitute the TokenizerFiles struct)
+let custom_model = TextEmbedding::try_new_from_user_defined(UserDefinedEmbeddingModel, options);
+
```
This should be a separate snippet probably.
Since it's not straightforward how a user can instantiate `UserDefinedEmbeddingModel` and that `read_file_to_bytes()` exists.

jcorrie replied:
Sorry, what do you mean by separate snippet? 

I'll add some visibility for the read_file_to_bytes function. (Just FYI, I wasn't intending for that to be super exposed because it's not really the job of this library, but agree it would be better DX.)

Anush008 replied:
Like a separate section in the README.md with a code example to load custom models.

Examples are the best docs, I believe.

jcorrie replied:
Understood - thanks. I'll submit over the next few days.

Anush008 commented:
HI @jcorrie.
We'll have to do a release to introduce the #44 fix. I'll include this PR too.
We can work on the README.md later.

jcorrie commented:
> HI @jcorrie. We'll have to do a release to introduce the #44 fix. I'll include this PR too. We can work on the README.md later.

It looks like this is all solved? Nothing for me to do other than the readme?
